test/e2e/declarative_test_fwk/helper.py

```diff
@@ -0,0 +1,261 @@
+# Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You may
+# not use this file except in compliance with the License. A copy of the
+# License is located at
+#
+#	 http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is distributed
+# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+
+"""Helper for Declarative tests framework for custom resources
+"""
+
+from e2e.declarative_test_fwk import model
+
+import logging
+from typing import Tuple
+from time import sleep
+from acktest.k8s import resource as k8s
+
+# holds custom resource helper references
+TEST_HELPERS = dict()
+
+
+def register_resource_helper(resource_kind: str, resource_plural: str):
+    """Decorator to discover Custom Resource Helper
+
+    Args:
+        resource_kind: custom resource kind
+        resource_plural: custom resource kind plural
+
+    Returns:
+        wrapper
+    """
+
+    def registrar(cls):
+        global TEST_HELPERS
+        if issubclass(cls, ResourceHelper):
+            TEST_HELPERS[resource_kind.lower()] = cls
+            cls.resource_plural = resource_plural.lower()
+            logging.info(f"Registered ResourceHelper: {cls.__name__} for custom resource kind: {resource_kind}")
+        else:
+            msg = f"Unable to register helper for {resource_kind} resource: {cls} is not a subclass of ResourceHelper"
+            logging.error(msg)
+            raise Exception(msg)
+    return registrar
+
+
+class ResourceHelper:
+    """Provides generic verb (create, patch, delete) methods for custom resources.
+    Keep its methods stateless. Methods are on instance to allow specialization.
+    """
+
+    DEFAULT_WAIT_SECS = 30
+
+    def create(self, input_data: dict, input_replacements: dict = {}) -> Tuple[k8s.CustomResourceReference, dict]:
+        """Creates custom resource inside Kubernetes cluster per the specifications in input data.
+
+        Args:
+            input_data: custom resource details
+            input_replacements: input replacements
+
+        Returns:
+            k8s.CustomResourceReference, created custom resource
+        """
+
+        reference = self.custom_resource_reference(input_data, input_replacements)
+        _ = k8s.create_custom_resource(reference, input_data)
+        resource = k8s.wait_resource_consumed_by_controller(reference, wait_periods=10)
+        assert resource is not None
```

Comment:
Is it OK to use `assert` statements in non unit test files? I'm not very experienced with python but i always thought that it's discouraged to use `assert` in libraries/functions that be imported by other projects.

Reply:
As the file is used in a test context I used the assert so that it helps in debug if it fails.
As we use this framework in e2e tests, we can evolve the code to keep/remove it later.